{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "09084a7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac8c08be",
   "metadata": {},
   "source": [
    "Burda işleme başlamadan önce eğer kolon adlarından önce idler yazıyorsa sadece kolon isimleri kalacak şekilde silin o fazlalıkları."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "49f794f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DEĞİŞTİR\n",
    "file1_path='MG25-3330.xlsx'\n",
    "file2_path='MG25-3331.xlsx'\n",
    "file3_path='MG25-3332.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "69bd5fc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "data1=pd.read_excel(file1_path)\n",
    "data2=pd.read_excel(file2_path)\n",
    "data3=pd.read_excel(file3_path)\n",
    "\n",
    "string1=re.findall(pattern=r'MG25-(\\d+)',string=file1_path)[0]\n",
    "string2=re.findall(pattern=r'MG25-(\\d+)',string=file2_path)[0] \n",
    "string3=re.findall(pattern=r'MG25-(\\d+)',string=file3_path)[0] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dbe06fb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_duo=pd.merge(data1,data2,on=['Chromosome','Position','End Position','Reference Allele','Sample Allele'],how='outer',suffixes=('*'+string1,'*'+string2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dfca94b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols=[a for a in data3.columns if a not in  ['Chromosome', 'Position', 'End Position', 'Reference Allele', 'Sample Allele']] \n",
    "\n",
    "a='*'+string3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "196fa241",
   "metadata": {},
   "outputs": [],
   "source": [
    "data3.columns = ['{}{}'.format(c, str(a) if c in cols else '') for c in data3.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "59e569f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Duo_Merge(file1_path, file2_path,output_filename): \n",
    "    data1=pd.read_excel(file1_path)\n",
    "    data2=pd.read_excel(file2_path)\n",
    "    string1=re.findall(pattern=r'_(\\d+)\\.xlsx',string=file1_path)[0]\n",
    "    string2=re.findall(pattern=r'_(\\d+)\\.xlsx',string=file2_path)[0] \n",
    "    data_duo=pd.merge(data1,data2,on=['Chromosome','Position','End Position','Reference Allele','Sample Allele'],how='outer',suffixes=('*'+string1,'*'+string2))\n",
    "    cols=sorted(data_duo.columns,key=lambda x: data1.columns.get_loc(x.split('*')[0]))\n",
    "    data_duo_update=data_duo[cols]\n",
    "    data_duo_update.columns= data_duo_update.columns.str.replace(r\"*\",\"_\")\n",
    "    data_duo_update.to_excel(output_filename,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f38e6d78",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_trio=pd.merge(data_duo,data3,on=['Chromosome','Position','End Position','Reference Allele','Sample Allele'],how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1cb5dfcb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\esra.bardakci\\AppData\\Local\\Temp\\ipykernel_4592\\2475845572.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  data_trio_update.columns= data_trio_update.columns.str.replace(r\"*\",\"_\")\n"
     ]
    }
   ],
   "source": [
    "col=sorted(data_trio.columns,key=lambda x: data1.columns.get_loc(x.split('*')[0]))\n",
    "\n",
    "data_trio_update=data_trio[col]\n",
    "\n",
    "data_trio_update.columns= data_trio_update.columns.str.replace(r\"*\",\"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6156cb25",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DEĞİŞTİR\n",
    "output_file='MG25-3330-3331-3332.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a937bb87",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_trio_update.to_excel(output_file,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61da1009",
   "metadata": {},
   "source": [
    "----------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "249cb887",
   "metadata": {},
   "outputs": [],
   "source": [
    "data=pd.read_excel(output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "33ef0278",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.drop([f'Conservation phyloP p-value_{string1}',f'Conservation phyloP p-value_{string2}',f'Conservation phyloP p-value_{string3}',f'Regulatory Site_{string1}',f'Regulatory Site_{string2}',f'Regulatory Site_{string3}',\n",
    "                  f'Regulator_{string1}',f'Regulator_{string2}',f'Regulator_{string3}',f'Transcriptional Regulator_{string1}',f'Transcriptional Regulator_{string2}',f'Transcriptional Regulator_{string3}',\n",
    "                  f'TFBS Enhancer Binding Score_{string1}',f'TFBS Enhancer Binding Score_{string2}',f'TFBS Enhancer Binding Score_{string3}',f'Enhancer Binding Site_{string1}',f'Enhancer Binding Site_{string2}',f'Enhancer Binding Site_{string3}',\n",
    "                 f'microRNA Regulator_{string1}',f'microRNA Regulator_{string2}',f'microRNA Regulator_{string3}',f'Gene Fusion_{string1}',f'Gene Fusion_{string2}',f'Gene Fusion_{string3}',f'Fused to_{string1}',f'Fused to_{string2}',f'Fused to_{string3}',\n",
    "                 f'DGV ID_{string1}',f'DGV ID_{string2}',f'DGV ID_{string3}',f'Odds Ratio_{string1}',f'Odds Ratio_{string2}',f'Odds Ratio_{string3}',f'Association p-value_{string1}',f'Association p-value_{string2}',f'Association p-value_{string3}',\n",
    "                 f'AFC Frequency_{string1}',f'AFC Frequency_{string2}',f'AFC Frequency_{string3}',f'CLINVAR ID_{string1}',f'CLINVAR ID_{string2}',f'CLINVAR ID_{string3}',f'CentoMD_{string1}',f'CentoMD_{string2}',f'CentoMD_{string3}',f'PGMD_{string1}',f'PGMD_{string2}',f'PGMD_{string3}',\n",
    "                 f'COSMIC ID_{string1}',f'COSMIC ID_{string2}',f'COSMIC ID_{string3}',f'Cosmic Frequency_{string1}',f'Cosmic Frequency_{string2}',f'Cosmic Frequency_{string3}', f'SIFT Function Prediction_{string1}',f'SIFT Function Prediction_{string2}',f'SIFT Function Prediction_{string3}',\n",
    "                  f'PolyPhen-2 Function Prediction_{string1}',f'PolyPhen-2 Function Prediction_{string2}',f'PolyPhen-2 Function Prediction_{string3}',\n",
    "               f'gnomAD Frequency_{string1}',f'gnomAD Frequency_{string2}',f'gnomAD Frequency_{string3}'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1f50c944",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_specified_columns(df, columns):\n",
    "    def merge_row(row):\n",
    "        values = [row[col] for col in columns if pd.notna(row[col]) and row[col] != '']\n",
    "        if not values:\n",
    "            return None\n",
    "        if all(value == values[0] for value in values):\n",
    "            return values[0]\n",
    "        return values[0]  # If not identical, return the first non-empty value (or define another rule if needed)\n",
    "    \n",
    "    return df.apply(merge_row, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "871f3aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the columns to merge\n",
    "columns_to_merge = [ f'Lab_frequency_{string1}', f'Lab_frequency_{string2}', f'Lab_frequency_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Lab_frequency'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'n_{string1}', f'n_{string2}',f'n_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['n'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'OMIM ID_{string1}', f'OMIM ID_{string2}',f'OMIM ID_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['OMIM ID'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'OMIM Phenotype_{string1}', f'OMIM Phenotype_{string2}',f'OMIM Phenotype_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['OMIM Phenotype'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'OMIM Gene_{string1}', f'OMIM Gene_{string2}',f'OMIM Gene_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['OMIM Gene'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'HGMD_{string1}',f'HGMD_{string2}',f'HGMD_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['HGMD'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'dbSNP ID_{string1}',f'dbSNP ID_{string2}',f'dbSNP ID_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['dbSNP ID'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'CADD Score_{string1}',f'CADD Score_{string2}',f'CADD Score_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['CADD Score'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'SIFT Score_{string1}',f'SIFT Score_{string2}',f'SIFT Score_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['SIFT Score'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Classification_{string1}',f'Classification_{string2}',f'Classification_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Classification'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Translation Impact_{string1}', f'Translation Impact_{string2}',f'Translation Impact_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Translation Impact'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Sample Upstream Filtering_{string1}', f'Sample Upstream Filtering_{string2}',f'Sample Upstream Filtering_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Sample Upstream Filtering'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Copy Number_{string1}', f'Copy Number_{string2}',f'Copy Number_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Copy Number'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Compound Heterozygous_{string1}',f'Compound Heterozygous_{string2}',f'Compound Heterozygous_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Compound Heterozygous'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Inferred Activity_{string1}', f'Inferred Activity_{string2}',f'Inferred Activity_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Inferred Activity'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Case Samples With Variant_{string1}',f'Case Samples With Variant_{string2}',f'Case Samples With Variant_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Case Samples With Variant'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Variant Findings_{string1}',f'Variant Findings_{string2}',f'Variant Findings_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Variant Findings'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Protein Variant_{string1}',f'Protein Variant_{string2}',f'Protein Variant_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Protein Variant'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Transcript Variant_{string1}',f'Transcript Variant_{string2}',f'Transcript Variant_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Transcript Variant'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Transcript ID_{string1}',f'Transcript ID_{string2}',f'Transcript ID_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Transcript ID'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Gene Symbol_{string1}',f'Gene Symbol_{string2}',f'Gene Symbol_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Gene Symbol'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Gene Region_{string1}',f'Gene Region_{string2}',f'Gene Region_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Gene Region'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Variation Type_{string1}',f'Variation Type_{string2}',f'Variation Type_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Variation Type'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Cytoband_{string1}',f'Cytoband_{string2}',f'Cytoband_{string3}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Cytoband'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f0c47d73",
   "metadata": {},
   "outputs": [],
   "source": [
    "data2=data[['Chromosome','Position','End Position','Cytoband','Reference Allele', 'Sample Allele','Variation Type','Gene Region',\n",
    "           'Gene Symbol','Transcript ID','Transcript Variant','Protein Variant','Variant Findings',f'Case Samples_{string1}',f'Case Samples_{string2}',f'Case Samples_{string3}','Case Samples With Variant',\n",
    "           'Inferred Activity',f'Genotype_{string1}',f'Genotype_{string2}',f'Genotype_{string3}',f'Sample Genotype Quality_{string1}',f'Sample Genotype Quality_{string2}',f'Sample Genotype Quality_{string3}','Compound Heterozygous',f'Call Quality_{string1}',f'Call Quality_{string2}',f'Call Quality_{string3}',\n",
    "            f'Read Depth_{string1}',f'Read Depth_{string2}',f'Read Depth_{string3}',f'Allele Fraction_{string1}',f'Allele Fraction_{string2}',f'Allele Fraction_{string3}',\n",
    "           'Copy Number','Sample Upstream Filtering','Translation Impact','Classification','SIFT Score','CADD Score','dbSNP ID','HGMD',\n",
    "           'OMIM Gene','OMIM Phenotype','OMIM ID','n','Lab_frequency']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e3b960d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DEĞİŞTİR\n",
    "data2.to_excel('MG25-3330-3331-3332-WGS-TRIO-hg38.xlsx',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bae0874",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
