{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a7002402",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a51649b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DEĞİŞTİR\n",
    "index1 = \"MG25-4143\"\n",
    "index2 = \"MG25-4144\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fe4a56e",
   "metadata": {},
   "source": [
    "Duo_Merge function takes 2 input files and output file name and writes merged version of two input files in a excel file which is named as output file name given at the beginning."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "25040ae2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Duo_Merge(file1_path, file2_path,output_filename, index1, index2): \n",
    "    data1=pd.read_excel(file1_path)\n",
    "    data2=pd.read_excel(file2_path)\n",
    "    #string1='MG25-3947'\n",
    "    #string2='MG25-3948'\n",
    "    data_duo=pd.merge(data1,data2,on=['Chromosome','Position','End Position','Reference Allele','Sample Allele'],how='outer',suffixes=('*'+index1,'*'+index2))\n",
    "    cols=sorted(data_duo.columns,key=lambda x: data1.columns.get_loc(x.split('*')[0]))\n",
    "    data_duo_update=data_duo[cols]\n",
    "    data_duo_update.columns= data_duo_update.columns.str.replace(r\"*\",\"_\")\n",
    "    data_duo_update.to_excel(output_filename,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5c5d4dbb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\esra.bardakci\\AppData\\Local\\Temp\\ipykernel_17208\\1039096735.py:9: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.\n",
      "  data_duo_update.columns= data_duo_update.columns.str.replace(r\"*\",\"_\")\n"
     ]
    }
   ],
   "source": [
    "#DEĞİŞTİR\n",
    "Duo_Merge('MG25-4143.xlsx','MG25-4144.xlsx','MG25-4143-4144-WGS-DUO.xlsx', index1, index2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "661fb82b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DEĞİŞTİR\n",
    "data=pd.read_excel('MG25-4143-4144-WGS-DUO.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ef24abfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.drop([f'Conservation phyloP p-value_{index1}', f'Conservation phyloP p-value_{index2}',f'Regulatory Site_{index1}',f'Regulatory Site_{index2}',\n",
    "                  f'Regulator_{index1}',f'Regulator_{index2}',f'Transcriptional Regulator_{index1}',f'Transcriptional Regulator_{index2}',\n",
    "                  f'TFBS Enhancer Binding Score_{index1}',f'TFBS Enhancer Binding Score_{index2}',f'Enhancer Binding Site_{index1}',f'Enhancer Binding Site_{index2}',\n",
    "                 f'microRNA Regulator_{index1}',f'microRNA Regulator_{index2}',f'Gene Fusion_{index1}',f'Gene Fusion_{index2}',f'Fused to_{index1}',f'Fused to_{index2}',\n",
    "                 f'DGV ID_{index1}',f'DGV ID_{index2}',f'Odds Ratio_{index1}',f'Odds Ratio_{index2}',f'Association p-value_{index1}',f'Association p-value_{index2}',\n",
    "                 f'AFC Frequency_{index1}',f'AFC Frequency_{index2}',f'CLINVAR ID_{index1}',f'CLINVAR ID_{index2}',f'CentoMD_{index1}',f'CentoMD_{index2}',f'PGMD_{index1}',f'PGMD_{index2}',\n",
    "                 f'COSMIC ID_{index1}',f'COSMIC ID_{index2}',f'Cosmic Frequency_{index1}',f'Cosmic Frequency_{index2}',f'SIFT Function Prediction_{index1}',f'SIFT Function Prediction_{index2}',\n",
    "                  f'PolyPhen-2 Function Prediction_{index1}',f'PolyPhen-2 Function Prediction_{index2}',\n",
    "               f'gnomAD Frequency_{index1}',f'gnomAD Frequency_{index2}'],  axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8829ef42",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_specified_columns(df, columns):\n",
    "    def merge_row(row):\n",
    "        values = [row[col] for col in columns if pd.notna(row[col]) and row[col] != '']\n",
    "        if not values:\n",
    "            return None\n",
    "        if all(value == values[0] for value in values):\n",
    "            return values[0]\n",
    "        return values[0]  # If not identical, return the first non-empty value (or define another rule if needed)\n",
    "    \n",
    "    return df.apply(merge_row, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "513adff5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Lab_frequency_{index1}', f'Lab_frequency_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Lab_frequency'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'n_{index1}', f'n_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['n'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'OMIM ID_{index1}', f'OMIM ID_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['OMIM ID'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'OMIM Phenotype_{index1}', f'OMIM Phenotype_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['OMIM Phenotype'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'OMIM Gene_{index1}', f'OMIM Gene_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['OMIM Gene'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'HGMD_{index1}',f'HGMD_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['HGMD'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'dbSNP ID_{index1}',f'dbSNP ID_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['dbSNP ID'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'CADD Score_{index1}',f'CADD Score_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['CADD Score'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'SIFT Score_{index1}',f'SIFT Score_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['SIFT Score'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Classification_{index1}',f'Classification_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Classification'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Translation Impact_{index1}', f'Translation Impact_{index2}',]\n",
    "\n",
    "# Create the merged column\n",
    "data['Translation Impact'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Sample Upstream Filtering_{index1}', f'Sample Upstream Filtering_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Sample Upstream Filtering'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Copy Number_{index1}', f'Copy Number_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Copy Number'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Compound Heterozygous_{index1}',f'Compound Heterozygous_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Compound Heterozygous'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [ f'Inferred Activity_{index1}',  f'Inferred Activity_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Inferred Activity'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Case Samples With Variant_{index1}', f'Case Samples With Variant_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Case Samples With Variant'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Variant Findings_{index1}',f'Variant Findings_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Variant Findings'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Protein Variant_{index1}',f'Protein Variant_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Protein Variant'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Transcript Variant_{index1}',f'Transcript Variant_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Transcript Variant'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Transcript ID_{index1}',f'Transcript ID_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Transcript ID'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Gene Symbol_{index1}',f'Gene Symbol_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Gene Symbol'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Gene Region_{index1}',f'Gene Region_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Gene Region'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Variation Type_{index1}',f'Variation Type_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Variation Type'] = merge_specified_columns(data, columns_to_merge)\n",
    "\n",
    "#------------------------------------------------------\n",
    "\n",
    "# Specify the columns to merge\n",
    "columns_to_merge = [f'Cytoband_{index1}',f'Cytoband_{index2}']\n",
    "\n",
    "# Create the merged column\n",
    "data['Cytoband'] = merge_specified_columns(data, columns_to_merge)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "91c6a650",
   "metadata": {},
   "outputs": [],
   "source": [
    "data2=data[['Chromosome','Position','End Position','Cytoband','Reference Allele', 'Sample Allele','Variation Type','Gene Region',\n",
    "           'Gene Symbol','Transcript ID','Transcript Variant','Protein Variant','Variant Findings',f'Case Samples_{index1}',f'Case Samples_{index2}','Case Samples With Variant',\n",
    "           'Inferred Activity',f'Genotype_{index1}',f'Genotype_{index2}',f'Sample Genotype Quality_{index1}',f'Sample Genotype Quality_{index2}','Compound Heterozygous',f'Call Quality_{index1}',f'Call Quality_{index2}',\n",
    "            f'Read Depth_{index1}',f'Read Depth_{index2}',f'Allele Fraction_{index1}',f'Allele Fraction_{index2}',\n",
    "           'Copy Number','Sample Upstream Filtering','Translation Impact','Classification','SIFT Score','CADD Score','dbSNP ID','HGMD',\n",
    "           'OMIM Gene','OMIM Phenotype','OMIM ID','n','Lab_frequency']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cf745669",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DEĞİŞTİR\n",
    "data2.to_excel(\"MG25-4143-4144-WGS-DUO-hg38-final.xlsx\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eea8fa0f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cab3453",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
